Label /etc/resolv.conf, /etc/hosts ... correctly in shared network co…

rhatdan opened the pull request:
…ntainers

If I run two containers with the same network they share the same /etc/resolv.conf.
The current code changes the labels of the /etc/resolv.conf currently to the
private label which causes it to be unusable in the first container.

This patch changes the labels to a shared label if more then one container
will use the content.

Docker-DCO-1.1-Signed-off-by: Dan Walsh dwalsh@redhat.com (github: rhatdan)

rhatdan commented:
Reported in this bugzilla.

rhatdan commented:
@jfrazelle @crosbymichael @tianon This should go into docker-1.8  It is a bug fix. 

Currently on SELinux systems using --net=container:UUID

SELinux will blow up the second container, making the resolver unreadable.  Common use case for kubernetes.

Fairly small SELinux change.

thaJeztah commented:
ping @calavera (he's the release captain for 1.8)

thaJeztah commented:
@rhatdan looks like the AUFS cleanup changes her are not related? (I may be wrong), perhaps a separate PR?

thaJeztah commented:
LOL, looks like you and Jess were merged as well 😹

    Docker-DCO-1.1-Signed-off-by: Jessie Frazelle <jfrazelle@users.noreply.github.com> (github: rhatdan)

calavera commented on daemon/container_unix.go:
```diff
@@ -1161,8 +1161,12 @@ func (container *Container) CleanupStorage() error {
 	var mounts []execdriver.Mount
+	mode := "Z"
+	if container.hostConfig.NetworkMode.IsContainer() {
+		mode = "z                                             "
```
are all those white spaces necessary?

rhatdan replied:
On 07/28/2015 08:23 PM, David Calavera wrote:
>
> In daemon/container_unix.go
> <https://github.com/docker/docker/pull/14922#discussion_r35717352>:
>
> > @@ -1161,8 +1161,12 @@ func (container *Container) CleanupStorage() error {
> >  
> >  func (container *Container) networkMounts() []execdriver.Mount {
> >  	var mounts []execdriver.Mount
> > +	mode := "Z"
> > +	if container.hostConfig.NetworkMode.IsContainer() {
> > +		mode = "z                                             "
>
> are all those white spaces necessary?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/14922/files#r35717352>.
>
Nope sorry.

calavera commented:
Can we remove the AUFS changes from this PR?

SvenDowideit commented:
is there no way to write an automated test for this?

SvenDowideit commented:
on my RHEL7 box, I get the following failures when running `make test`:

```
----------------------------------------------------------------------
FAIL: docker_cli_daemon_test.go:62: DockerDaemonSuite.TestDaemonRestartWithVolumesRefs

waiting for daemon to start
daemon started
exiting daemon
waiting for daemon to start
daemon started
docker_cli_daemon_test.go:75:
    c.Fatal(err)
... Error: exit status 1

exiting daemon

----------------------------------------------------------------------
FAIL: docker_cli_daemon_test.go:99: DockerDaemonSuite.TestDaemonStartBridgeWithoutIPAssociation

waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
docker_cli_daemon_test.go:104:
    c.Fatalf("Could not start daemon: %v", err)
... Error: Could not start daemon: Daemon exited and never started


----------------------------------------------------------------------
FAIL: docker_cli_daemon_test.go:90: DockerDaemonSuite.TestDaemonStartIptablesFalse

waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
docker_cli_daemon_test.go:92:
    c.Fatalf("we should have been able to start the daemon with passing iptables=false: %v", err)
... Error: we should have been able to start the daemon with passing iptables=false: Daemon exited and never started


----------------------------------------------------------------------
FAIL: docker_cli_daemon_test.go:829: DockerDaemonSuite.TestDaemonUlimitDefaults

waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
waiting for daemon to start
docker_cli_daemon_test.go:833:
    c.Fatal(err)
... Error: Daemon exited and never started


----------------------------------------------------------------------
PASS: docker_cli_daemon_test.go:1091: DockerDaemonSuite.TestDaemonUnixSockCleanedUp     0.503s
PASS: docker_cli_daemon_test.go:1116: DockerDaemonSuite.TestDaemonWithWrongkey  6.505s

----------------------------------------------------------------------
FAIL: docker_cli_exec_test.go:90: DockerDaemonSuite.TestExecAfterDaemonRestart

waiting for daemon to start
daemon started
docker_cli_exec_test.go:94:
    c.Fatalf("Could not start daemon with busybox: %v", err)
... Error: Could not start daemon with busybox: could not save busybox image: exit status 1

exiting daemon

----------------------------------------------------------------------
FAIL: check_test.go:56: DockerDaemonSuite.SetUpTest

check_test.go:57:
    s.d = NewDaemon(c)
github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go:79:
    ...open github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go: no such file or directory
... Error: Could not create /go/src/github.com/docker/docker/bundles/1.8.0-dev/test-integration-cli/d9908769/graph directory


----------------------------------------------------------------------
PANIC: docker_cli_daemon_test.go:1211: DockerDaemonSuite.TestHttpsInfo

... Panic: Fixture has panicked (see related PANIC)

----------------------------------------------------------------------
MISS: docker_cli_daemon_test.go:1232: DockerDaemonSuite.TestHttpsInfoRogueCert
MISS: docker_cli_daemon_test.go:1253: DockerDaemonSuite.TestHttpsInfoRogueServerCert
MISS: docker_cli_daemon_test.go:1319: DockerDaemonSuite.TestRunContainerWithBridgeNone

----------------------------------------------------------------------
FAIL: docker_cli_start_volume_driver_unix_test.go:40: DockerExternalVolumeSuite.SetUpTest

docker_cli_start_volume_driver_unix_test.go:41:
    s.d = NewDaemon(c)
github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go:79:
    ...open github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go: no such file or directory
... Error: Could not create /go/src/github.com/docker/docker/bundles/1.8.0-dev/test-integration-cli/d47129329/graph directory


----------------------------------------------------------------------
PANIC: docker_cli_start_volume_driver_unix_test.go:45: DockerExternalVolumeSuite.TearDownTest

... Panic: runtime error: invalid memory address or nil pointer dereference (PC=0x43D9C5)

/usr/local/go/src/runtime/panic.go:387
  in gopanic
/usr/local/go/src/runtime/panic.go:42
  in panicmem
/usr/local/go/src/runtime/sigpanic_unix.go:26
  in sigpanic
github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go:275
  in Daemon.Stop
docker_cli_start_volume_driver_unix_test.go:46
  in DockerExternalVolumeSuite.TearDownTest
/usr/local/go/src/reflect/value.go:296
  in Value.Call
/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:712
  in func.002
/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:657
  in func.001
/usr/local/go/src/runtime/asm_amd64.s:2232
  in goexit

----------------------------------------------------------------------
PANIC: docker_cli_start_volume_driver_unix_test.go:148: DockerExternalVolumeSuite.TestStartExternalNamedVolumeDriver

... Panic: Fixture has panicked (see related PANIC)

----------------------------------------------------------------------
MISS: <autogenerated>:30: DockerExternalVolumeSuite.TestStartExternalVolumeDriverDeleteContainer
MISS: <autogenerated>:29: DockerExternalVolumeSuite.TestStartExternalVolumeDriverVolumesFrom
MISS: docker_cli_start_volume_driver_unix_test.go:179: DockerExternalVolumeSuite.TestStartExternalVolumeUnnamedDriver
OOPS: 774 passed, 6 skipped, 9 FAILED, 1 FIXTURE-PANICKED, 8 MISSED
--- FAIL: Test (954.25s)
FAIL
testing: open /go/src/github.com/docker/docker/bundles/1.8.0-dev/test-integration-cli/coverprofiles/docker-integration-cli: no space left on device
exit status 2
FAIL    github.com/docker/docker/integration-cli        954.302s
---> Making bundle: .integration-daemon-stop (in bundles/1.8.0-dev/test-integration-cli)
+++++ cat bundles/1.8.0-dev/test-integration-cli/docker.pid
++++ kill 9156
```

rhatdan committed:
Labels on network content need to be shared if shared network namespace

If I run two containers with the same network they share the same /etc/resolv.conf.
The current code changes the labels of the /etc/resolv.conf currently to the
private label which causes it to be unusable in the first container.

This patch changes the labels to a shared label if more then one container
will use the content.

Docker-DCO-1.1-Signed-off-by: Dan Walsh dwalsh@redhat.com (github: rhatdan)

Docker-DCO-1.1-Signed-off-by: Dan Walsh <dwalsh@redhat.com> (github: rhatdan)

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "netlabel" git@github.com:rhatdan/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

rhatdan commented:
Sorry about the dirty patch.  I have cleaned it up.  Looks like Gordon does not like the way I sign my content any longer?

rhatdan commented:
@SvenDowideit None of those errors are caused by this patch, I believe.

thaJeztah commented:
possibly the missing "pointy brackets" around your email

But it's easier to just use `git commit -s`, that should add a proper sign-off

rhatdan commented:
@SvenDowideit Do we have any tests for sharing networks between containers?

rhatdan commented:
We could look at the label of /etc/resolv.conf when a container starts and then look at it when a second container joins the network, which should force the label to end in `:s0`

rhatdan commented:
@thaJeztah Some times it likes it sometimes it doesn't, I have git automatically adding the DSO.

calavera commented:
thanks @rhatdan. Yeah, CI is kinda broken right now, nothing to worry about just jet. Is there any way we could have a test for this change?

the changes LGTM, but just in case.

rhatdan commented:
@calavera Do you know of any tests that look at sharing namespaces between containers, that I could use as an example?

calavera commented:
unfortunately, I don't.

/cc @LK4D4, @icecrime

icecrime commented:
@rhatdan I see a lot of tests involving `--net=container:<name>` in `integration-cli/docker_cli_run_unix_test.go`, is that what you were looking for? Examples: `TestRunContainerNetModeWithDnsMacHosts`, `TestRunContainerNetModeWithExposePort`, `TestRunModeNetContainerHostname`, ...

icecrime commented:
A test would be nice, otherwise code LGTM.

rhatdan commented:
@icecrime That is what I was looking for thanks.  I will try to throw something together if I get time.

calavera commented:
Merging this as it is. It's ready for the release.

calavera commented:
cherry picked into #15091

thaJeztah commented:
Do we still want tests (as a separate PR)?

calavera commented:
@thaJeztah we don't have a way to test SElinux changes in a reliable way at the moment.

calavera commented:
I tested it manually though.

thaJeztah commented:
@calavera bummer. Guess it's okay then (I like tests to detect regressions). Thanks!

rhatdan commented:
As long as you are testing within a container, SELinux will be disabled, so nothing will happen.  If you run a test outside a container, we could write tests.

thaJeztah commented:
@jfrazelle do you think it's worth exploring? Perhaps conditional tests (`if !runningIncontainer { ... }`)
